File: python/day24.py
import unittest
from collections import defaultdict
from typing import List, Set, Tuple

from utilities import read_lines

direction_moves = {
    'e': (1, 0),
    'w': (-1, 0),
    'ne': (0, 1),
    'sw': (0, -1),
    'nw': (-1, 1),
    'se': (1, -1),
}

directions = ['ne', 'nw', 'se', 'sw', 'w', 'e']


def solve_part_1(lines: List[str]) -> Set[Tuple[int, int]]:
    tiles = set()
    for line in lines:
        pos = (0, 0)
        i = 0
        while i < len(line):
            for direction in directions:
                if line[i:i+len(direction)] == direction:
                    move = direction_moves[direction]
                    pos = (pos[0] + move[0], pos[1] + move[1])
                    i += len(direction)
                    break
        if pos in tiles:
            tiles.remove(pos)
        else:
            tiles.add(pos)
    return tiles


def solve_part_2(tiles) -> int:
    for _ in range(100):
        no_flip = set()
        whites = defaultdict(int)
        for pos in tiles:
            count = 0
            for move in direction_moves.values():
                new_pos = (pos[0] + move[0], pos[1] + move[1])
                if new_pos in tiles:
                    # Black neighbour of the current pos
                    count += 1
                else:
                    # White neighbour of the current pos, so increase
                    # the number of black neighbour for this white neighbour
                    # by 1.
                    whites[new_pos] += 1

            if count == 1 or count == 2:
                # This black tile will not be flipped
                no_flip.add(pos)
        white_to_black = set(
            [pos for pos, black_neighbour_count in whites.items() if black_neighbour_count == 2])

        tiles = no_flip.union(white_to_black)
    return len(tiles)


class Testing(unittest.TestCase):

    test_input = """sesenwnenenewseeswwswswwnenewsewsw
neeenesenwnwwswnenewnwwsewnenwseswesw
seswneswswsenwwnwse
nwnwneseeswswnenewneswwnewseswneseene
swweswneswnenwsewnwneneseenw
eesenwseswswnenwswnwnwsewwnwsene
sewnenenenesenwsewnenwwwse
wenwwweseeeweswwwnwwe
wsweesenenewnwwnwsenewsenwwsesesenwne
neeswseenwwswnwswswnw
nenwswwsewswnenenewsenwsenwnesesenew
enewnwewneswsewnwswenweswnenwsenwsw
sweneswneswneneenwnewenewwneswswnese
swwesenesewenwneswnwwneseswwne
enesenwswwswneneswsenwnewswseenwsese
wnwnesenesenenwwnenwsewesewsesesew
nenewswnwewswnenesenwnesewesw
eneswnwswnwsenenwnwnwwseeswneewsenese
neswnwewnwnwseenwseesewsenwsweewe
wseweeenwnesenwwwswnew"""

    def test_part_1(self):
        tiles = solve_part_1(self.test_input.splitlines())
        self.assertEqual(len(tiles), 10)
        self.assertEqual(solve_part_2(tiles), 2208)


if __name__ == '__main__':
    print('Day 24')
    import time
    start = time.time()
    lines = read_lines('../inputs/day24.txt')

    tiles = solve_part_1(lines)
    print('\tPart 1: {}'.format(len(tiles)))
    print('\tPart 2: {}'.format(solve_part_2(tiles)))

    print("\tTime: {:.4f}ms".format((time.time() - start) * 1000))
    unittest.main()

File: rust/src/day24.rs
use std::collections::{HashMap, HashSet};

use crate::utilities::read_lines;

const DIRECTIONS: [&str; 6] = ["ne", "nw", "se", "sw", "e", "w"];
const CHANGES: [(i16, i16); 6] = [(0, 1), (-1, 1), (1, -1), (0, -1), (1, 0), (-1, 0)];

fn solve_part_1(lines: Vec<String>) -> HashSet<(i16, i16)> {
    let mut black_tiles = HashSet::new();
    for line in lines {
        // Reference tile position
        let mut pos = (0, 0);
        let mut s = line.as_str();
        while !s.is_empty() {
            for (index, direction) in DIRECTIONS.iter().enumerate() {
                if s.starts_with(*direction) {
                    s = &s[direction.len()..];
                    pos.0 += CHANGES[index].0;
                    pos.1 += CHANGES[index].1;
                    break;
                }
            }
        }
        if black_tiles.contains(&pos) {
            black_tiles.remove(&pos);
        } else {
            black_tiles.insert(pos);
        }
    }
    black_tiles
}

fn solve_part_2(mut black_tiles: HashSet<(i16, i16)>) -> usize {
    for _ in 0..100 {
        let mut no_flip_black_tiles = HashSet::new();
        let mut white_tiles = HashMap::new();
        for tile in &black_tiles {
            let mut black_neighbors = 0;
            for change in &CHANGES {
                let neighbor = (tile.0 + change.0, tile.1 + change.1);
                if black_tiles.contains(&neighbor) {
                    black_neighbors += 1;
                } else {
                    // This is a white neighbor, so increase the number of
                    // black neighbors for this white neighbor by 1, because
                    // it's a neighbor to the current black `tile`
                    *white_tiles.entry(neighbor).or_insert(0) += 1;
                }
            }

            if black_neighbors == 1 || black_neighbors == 2 {
                no_flip_black_tiles.insert(*tile);
            }
        }
        let white_to_black: HashSet<(i16, i16)> = white_tiles
            .into_iter()
            .filter(|(_, black_neighbors)| *black_neighbors == 2)
            .map(|(pos, _)| pos)
            .collect();
        black_tiles = no_flip_black_tiles
            .union(&white_to_black)
            .cloned()
            .collect::<HashSet<(i16, i16)>>();
    }
    black_tiles.len()
}
pub fn main() {
    println!("Day 24");
    let lines = read_lines("../inputs/day24.txt");
    let black_tiles = solve_part_1(lines);
    println!("\tPart1: {}", black_tiles.len());
    println!("\tPart2: {}", solve_part_2(black_tiles));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_1() {
        let input_str = r#"sesenwnenenewseeswwswswwnenewsewsw
neeenesenwnwwswnenewnwwsewnenwseswesw
seswneswswsenwwnwse
nwnwneseeswswnenewneswwnewseswneseene
swweswneswnenwsewnwneneseenw
eesenwseswswnenwswnwnwsewwnwsene
sewnenenenesenwsewnenwwwse
wenwwweseeeweswwwnwwe
wsweesenenewnwwnwsenewsenwwsesesenwne
neeswseenwwswnwswswnw
nenwswwsewswnenenewsenwsenwnesesenew
enewnwewneswsewnwswenweswnenwsenwsw
sweneswneswneneenwnewenewwneswswnese
swwesenesewenwneswnwwneseswwne
enesenwswwswneneswsenwnewswseenwsese
wnwnesenesenenwwnenwsewesewsesesew
nenewswnwewswnenesenwnesewesw
eneswnwswnwsenenwnwnwwseeswneewsenese
neswnwewnwnwseenwseesewsenwsweewe
wseweeenwnesenwwwswnew"#;
        assert_eq!(
            solve_part_1(input_str.lines().map(str::to_string).collect()).len(),
            10
        );
    }

    #[test]
    fn test_part_2() {
        let input_str = r#"sesenwnenenewseeswwswswwnenewsewsw
neeenesenwnwwswnenewnwwsewnenwseswesw
seswneswswsenwwnwse
nwnwneseeswswnenewneswwnewseswneseene
swweswneswnenwsewnwneneseenw
eesenwseswswnenwswnwnwsewwnwsene
sewnenenenesenwsewnenwwwse
wenwwweseeeweswwwnwwe
wsweesenenewnwwnwsenewsenwwsesesenwne
neeswseenwwswnwswswnw
nenwswwsewswnenenewsenwsenwnesesenew
enewnwewneswsewnwswenweswnenwsenwsw
sweneswneswneneenwnewenewwneswswnese
swwesenesewenwneswnwwneseswwne
enesenwswwswneneswsenwnewswseenwsese
wnwnesenesenenwwnenwsewesewsesesew
nenewswnwewswnenesenwnesewesw
eneswnwswnwsenenwnwnwwseeswneewsenese
neswnwewnwnwseenwseesewsenwsweewe
wseweeenwnesenwwwswnew"#;
        let black_tiles = solve_part_1(input_str.lines().map(str::to_string).collect());
        assert_eq!(solve_part_2(black_tiles), 2208);
    }
}
